gnosis/safe-react
# feat: Add multi send option for queued transactions

## What it solves
Resolves #3687 

## How this PR fixes it
We are adding a _Batch Execute_ Button to the top of the Tx Queue. If there is more than one eligible transaction the user can opt to execute them with a single call to the MultiSend contract.

## ToDos

- [x] Handle Reject Transactions (multiple transactions for the same nonce)
- [x] Add Unit Tests for `BatchExecute` and `getBatchableTransactions`
- [x] Add Toggle to Settings page
- [x] Highlight batchable transactions on button hover
- [x] Add notifications on user interaction
- [x] Add tracking for Batch Execute Button
- [x] Add unit test for `BATCH_LIMIT`
- [x] 5 - The length of the Tx are of different width.
- [x] Show On-chain rejection in modals

## How to test it

This needs to be tested with different Safe versions and on different networks where the MultiSend Contract [might not be deployed](https://github.com/gnosis/safe-deployments/blob/main/src/assets/v1.1.1/multi_send.json) i.e. Avalanche with a v1.1.1 Safe.

1. Open the Safe App
2. Queue up multiple transactions
3. Make sure that their nonces are in order and fully sign them
4. Observe that the Button should become enabled as soon as there is more than one eligible transaction
5. Click the Button
6. Observe that a modal opens with more details
7. Click Submit
8. Observe that MM opens with the correct MultiSend call
9. Click Confirm in MM
10. Observe that a pending state is added to the transactions

## Analytics changes

Clicks on the `Execute Batch` Button are tracked with the number of transactions that are included in the batch.

## Screenshots
![Screenshot 2022-03-21 at 12 39 34](https://user-images.githubusercontent.com/5880855/159253967-6a777cc9-1fe2-4e66-8eb3-110461f9f424.png)
![Screenshot 2022-03-21 at 12 05 12](https://user-images.githubusercontent.com/5880855/159253377-8f4fdaa5-205f-4a90-b3c2-e24a98ae959e.png)
![Screenshot 2022-03-21 at 12 35 13](https://user-images.githubusercontent.com/5880855/159253382-6e87b509-d200-4724-be89-1d59acadf2c6.png)
![Screenshot 2022-03-23 at 15 37 02](https://user-images.githubusercontent.com/5880855/159724645-5348201a-315e-4eab-8338-7a324bde4841.png)


## Test Scenarios
Scenarios:
safe 1/1:
Batch tx with: 
- No txs  (not possible to excute in batch is expected)
- 1 tx nonces (not possible to excute in batch is expected)
- 2 tx nonces
- 2 tx nonces with more than 1 tx in the same nonce
- 2 tx nonces with one rejection
- 3 tx nonces
- 3 tx nonces with one of them guaranteed to fail
- tx executed by a non owner

Safe 2/2: v1.3.0 and 1.1.1
- Same as safe 1/1
- 2 tx nonces with only the 1st one fully signed
- 2 tx nonces with only the 2nd one fully signed (not possible to excute in batch is expected)
- 2 tx nonces. One of the nonces has 3 tx in parallel and the 1st one is fully signed 
- 2 tx nonces. One of the nonces has 3 tx in parallel and all of them are signed
- 2 tx nonces. One of the nonces has 3 tx in parallel with one of them a rejection, all fully signed
- 3 tx nonces with the first 2 fully signed
- 3 tx nonces with 1st and 3rd one fully signed (not possible to excute in batch is expected)
- 4 tx nonces with 1, 2 and 4 fully signed
